File: code/cpp/ch22/learn/example_02_monotonic_stack.cpp
/*
 * Example 02: Monotonic Stack — Next Greater Element & Histogram
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * Demonstrates:
 *   - Next Greater Element with monotonic stack
 *   - Largest Rectangle in Histogram
 *   - Sliding Window Maximum with deque
 *
 * Build: g++ -std=c++17 -o /tmp/ex02_ch22 code/cpp/ch22/learn/example_02_monotonic_stack.cpp && /tmp/ex02_ch22
 */

#include <algorithm>
#include <deque>
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

void nextGreaterDemo() {
    cout << "=== Next Greater Element ===" << endl;
    vector<int> arr = {4, 5, 2, 10, 8};
    int n = arr.size();
    vector<int> result(n, -1);
    stack<int> stk;

    for (int i = n - 1; i >= 0; i--) {
        while (!stk.empty() && arr[stk.top()] <= arr[i]) stk.pop();
        if (!stk.empty()) result[i] = arr[stk.top()];
        stk.push(i);
    }

    cout << "  Input:  ";
    for (int x : arr) cout << x << " ";
    cout << endl << "  Result: ";
    for (int x : result) cout << x << " ";
    cout << endl << endl;
}

void histogramDemo() {
    cout << "=== Largest Rectangle in Histogram ===" << endl;
    vector<int> heights = {2, 1, 5, 6, 2, 3};
    stack<int> stk;
    int maxArea = 0;
    int n = heights.size();

    for (int i = 0; i <= n; i++) {
        int curr = (i == n) ? 0 : heights[i];
        while (!stk.empty() && heights[stk.top()] > curr) {
            int h = heights[stk.top()]; stk.pop();
            int w = stk.empty() ? i : i - stk.top() - 1;
            maxArea = max(maxArea, h * w);
        }
        stk.push(i);
    }

    cout << "  Heights: ";
    for (int x : heights) cout << x << " ";
    cout << endl << "  Max area: " << maxArea << endl << endl;
}

void slidingWindowDemo() {
    cout << "=== Sliding Window Maximum ===" << endl;
    vector<int> nums = {1, 3, -1, -3, 5, 3, 6, 7};
    int k = 3;
    deque<int> dq;
    vector<int> result;

    for (int i = 0; i < (int)nums.size(); i++) {
        while (!dq.empty() && dq.front() < i - k + 1) dq.pop_front();
        while (!dq.empty() && nums[dq.back()] <= nums[i]) dq.pop_back();
        dq.push_back(i);
        if (i >= k - 1) result.push_back(nums[dq.front()]);
    }

    cout << "  Input: ";
    for (int x : nums) cout << x << " ";
    cout << ", k=" << k << endl << "  Result: ";
    for (int x : result) cout << x << " ";
    cout << endl;
}

int main() {
    nextGreaterDemo();
    histogramDemo();
    slidingWindowDemo();
    return 0;
}

File: code/java/ch22/learn/Example02MonotonicStack.java
package ch22.learn;

import java.util.*;

/**
 * Example 02: Monotonic Stack — Next Greater Element & Histogram
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * Demonstrates:
 *   - Next Greater Element using monotonic stack
 *   - Largest Rectangle in Histogram
 *   - Sliding Window Maximum with deque
 */
public class Example02MonotonicStack {

    public static void nextGreaterDemo() {
        System.out.println("=== Next Greater Element ===");
        int[] arr = {4, 5, 2, 10, 8};
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[stack.peek()] <= arr[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) result[i] = arr[stack.peek()];
            stack.push(i);
        }
        System.out.println("  Input:  " + Arrays.toString(arr));
        System.out.println("  Result: " + Arrays.toString(result));
        System.out.println();
    }

    public static void histogramDemo() {
        System.out.println("=== Largest Rectangle in Histogram ===");
        int[] heights = {2, 1, 5, 6, 2, 3};
        Deque<Integer> stack = new ArrayDeque<>();
        int maxArea = 0;
        int n = heights.length;

        for (int i = 0; i <= n; i++) {
            int curr = (i == n) ? 0 : heights[i];
            while (!stack.isEmpty() && heights[stack.peek()] > curr) {
                int h = heights[stack.pop()];
                int w = stack.isEmpty() ? i : i - stack.peek() - 1;
                int area = h * w;
                maxArea = Math.max(maxArea, area);
                System.out.println("  Pop h=" + h + ", w=" + w + ", area=" + area);
            }
            stack.push(i);
        }
        System.out.println("  Heights: " + Arrays.toString(heights));
        System.out.println("  Max area: " + maxArea);
        System.out.println();
    }

    public static void slidingWindowDemo() {
        System.out.println("=== Sliding Window Maximum ===");
        int[] nums = {1, 3, -1, -3, 5, 3, 6, 7};
        int k = 3;
        Deque<Integer> dq = new ArrayDeque<>();
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < nums.length; i++) {
            while (!dq.isEmpty() && dq.peekFirst() < i - k + 1) dq.pollFirst();
            while (!dq.isEmpty() && nums[dq.peekLast()] <= nums[i]) dq.pollLast();
            dq.offerLast(i);
            if (i >= k - 1) result.add(nums[dq.peekFirst()]);
        }
        System.out.println("  Input: " + Arrays.toString(nums) + ", k=" + k);
        System.out.println("  Result: " + result);
    }

    public static void main(String[] args) {
        nextGreaterDemo();
        histogramDemo();
        slidingWindowDemo();
    }
}
